pipeline/authn/authenticator_cookie_session.go

```diff
@@ -129,55 +151,67 @@ func cookieSessionResponsible(r *http.Request, only []string) bool {
 	return false
 }
 
-func forwardRequestToSessionStore(r *http.Request, checkSessionURL string, preserveQuery bool, preservePath bool, preserveHost bool, setHeaders map[string]string, m string) (json.RawMessage, error) {
-	reqUrl, err := url.Parse(checkSessionURL)
+func forwardRequestToSessionStore(r *http.Request, cf *AuthenticatorForwardConfig) (json.RawMessage, error) {
+	req, err := PrepareRequest(r, cf)
 	if err != nil {
-		return nil, errors.WithStack(herodot.ErrInternalServerError.WithReasonf("Unable to parse session check URL: %s", err))
+		return nil, err
 	}
 
-	if !preservePath {
-		reqUrl.Path = r.URL.Path
+	res, err := http.DefaultClient.Do(req.WithContext(r.Context()))
+	if err != nil {
+		return nil, helper.ErrForbidden.WithReason(err.Error()).WithTrace(err)
 	}
 
-	if !preserveQuery {
-		reqUrl.RawQuery = r.URL.RawQuery
-	}
+	defer res.Body.Close()
 
-	if m == "" {
-		m = r.Method
+	if res.StatusCode == http.StatusOK {
+		body, err := ioutil.ReadAll(res.Body)
+		if err != nil {
+			return json.RawMessage{}, errors.WithStack(herodot.ErrInternalServerError.WithReasonf("Unable to fetch cookie session context from remote: %+v", err))
+		}
+		return body, nil
+	} else {
+		return json.RawMessage{}, errors.WithStack(helper.ErrUnauthorized)
 	}
+}
 
-	req := http.Request{
-		Method: m,
-		URL:    reqUrl,
-		Header: http.Header{},
+func PrepareRequest(r *http.Request, cf *AuthenticatorForwardConfig) (http.Request, error) {
+	reqURL, err := url.Parse(cf.CheckSessionURL)
+	if err != nil {
+		return http.Request{}, errors.WithStack(herodot.ErrInternalServerError.WithReasonf("Unable to parse session check URL: %s", err))
 	}
 
-	// We need to make a COPY of the header, not modify r.Header!
-	for k, v := range r.Header {
-		req.Header[k] = v
+	if !cf.PreservePath {
+		reqURL.Path = r.URL.Path
 	}
 
-	for k, v := range setHeaders {
-		req.Header.Set(k, v)
+	if !cf.PreserveQuery {
+		reqURL.RawQuery = r.URL.RawQuery
 	}
 
-	if preserveHost {
-		req.Header.Set("X-Forwarded-Host", r.Host)
+	if cf.ForceMethod == "" {
+		cf.ForceMethod = r.Method
 	}
 
-	res, err := http.DefaultClient.Do(req.WithContext(r.Context()))
-	if err != nil {
-		return nil, helper.ErrForbidden.WithReason(err.Error()).WithTrace(err)
+	req := http.Request{
+		Method: cf.ForceMethod,
+		URL:    reqURL,
+		Header: http.Header{},
 	}
 
-	if res.StatusCode == 200 {
-		body, err := ioutil.ReadAll(res.Body)
-		if err != nil {
-			return json.RawMessage{}, errors.WithStack(herodot.ErrInternalServerError.WithReasonf("Unable to fetch cookie session context from remote: %+v", err))
+	// We need to copy only essential and configurable headers
+	for k, v := range r.Header {
+		if _, ok := cf.ForwardHTTPHeaders[k]; ok {
+			req.Header[k] = v
```

Comment:
My recommendation is to not use precomputed map here. You don't really save anything - it requires more memory (to store the map) and more CPU time (because of memory allocation). And it's easy to mess up since the "map" is sort of like a cache. It's totally fine to be explicit and have two loops:

```suggestion
	// We need to copy only essential and configurable headers
	for requested, v := range r.Header {
		for _, allowed := range cf.ForwardHTTPHeaders {
			if requested == allowed {
				req.Header[k] = v
```